pkg/deploy/install.go

```diff
@@ -65,39 +65,43 @@ func (deploy *Deployer) installNamespace() error {
-		if deploy.config.Platform == "openshift" {
-			if namespace.ObjectMeta.Labels != nil {
-				namespace.ObjectMeta.Labels["openshift.io/cluster-monitoring"] = "true"
-				deploy.logger.Infof("Updated the 'openshift.io/cluster-monitoring' label to the %s namespace", deploy.config.Namespace)
-			} else {
-				namespace.ObjectMeta.Labels = map[string]string{
-					"openshift.io/cluster-monitoring": "true",
-				}
-				deploy.logger.Infof("Added the 'openshift.io/cluster-monitoring' label to the %s namespace", deploy.config.Namespace)
-			}
-			if namespace.ObjectMeta.Annotations != nil {
-				namespace.ObjectMeta.Annotations["openshift.io/node-selector"] = ""
-				deploy.logger.Infof("Updated the 'openshift.io/node-selector' annotation to the %s namespace", deploy.config.Namespace)
-			} else {
-				namespace.ObjectMeta.Annotations = map[string]string{
-					"openshift.io/node-selector": "",
-				}
-				deploy.logger.Infof("Added the empty 'openshift.io/node-selector' annotation to the %s namespace", deploy.config.Namespace)
-			}
+		return nil
+	}
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		return err
+	}
 
-			_, err := deploy.client.CoreV1().Namespaces().Update(context.TODO(), namespace, metav1.UpdateOptions{})
-			if err != nil {
-				return fmt.Errorf("failed to add the 'openshift.io/cluster-monitoring' label to the %s namespace: %v", deploy.config.Namespace, err)
+	// check if we need to add/update the cluster-monitoring label for Openshift installs.
+	if deploy.config.Platform == "openshift" {
+		if namespace.ObjectMeta.Labels != nil {
+			namespace.ObjectMeta.Labels["openshift.io/cluster-monitoring"] = "true"
+			deploy.logger.Infof("Updated the 'openshift.io/cluster-monitoring' label to the %s namespace", deploy.config.Namespace)
+		} else {
+			namespace.ObjectMeta.Labels = map[string]string{
+				"openshift.io/cluster-monitoring": "true",
 			}
+			deploy.logger.Infof("Added the 'openshift.io/cluster-monitoring' label to the %s namespace", deploy.config.Namespace)
+		}
+		if namespace.ObjectMeta.Annotations != nil {
+			namespace.ObjectMeta.Annotations["openshift.io/node-selector"] = ""
+			deploy.logger.Infof("Updated the 'openshift.io/node-selector' annotation to the %s namespace", deploy.config.Namespace)
 		} else {
-			deploy.logger.Infof("The %s namespace already exists", deploy.config.Namespace)
+			namespace.ObjectMeta.Annotations = map[string]string{
+				"openshift.io/node-selector": "",
+			}
+			deploy.logger.Infof("Added the empty 'openshift.io/node-selector' annotation to the %s namespace", deploy.config.Namespace)
 		}
-	} else {
-		return err
+
+		_, err := deploy.client.CoreV1().Namespaces().Update(context.TODO(), namespace, metav1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to add the 'openshift.io/cluster-monitoring' label to the %s namespace: %v", deploy.config.Namespace, err)
+		}
+		return nil
 	}
 
+	// TODO: handle updating the namespace for non-openshift installations
```

Comment:
You can eliminate a few more of the redundancies and this TODO if you think about this method in two pieces: the ns itself and the labels/annotations.  Ie. you always treat the labels/annotations as an update.  

What you get: the updates for existing namespaces, removal of the duplicate code to annotate and label in the openshift deployment case, ensuring the `ExtraNamespaceLabels` are applied uniformly (looks like they're only applied here in the creation case)

What you lose: logging the add vs update but that could be done with some additional logic if it really is meaningful.  I guess the same for the labels/annotations cased for the openshift deployment type.  Does it harm anything to keep it on there in all cases? (Not looking for an answer, just something to consider)

So something like this: (note, I didn't try to write this in an IDE or check it for formatting, it may not even work.  Demo purposes only 😄 )


```go
func (deploy *Deployer) installNamespace() error {
	namespace, err := deploy.client.CoreV1().Namespaces().Get(context.TODO(), deploy.config.Namespace, metav1.GetOptions{})
	if err != nil && (!apierrors.IsNotFound(err) || !apierrors.IsAlreadyExists(err)) {
		return err
	}

	if apierrors.IsAlreadyExists(err) {
		deploy.logger.Infof("The %s namespace already exists", deploy.config.Namespace)
	}

	// not found - create the ns
	if apierrors.IsNotFound(err) {
		namespaceObjectMeta := metav1.ObjectMeta{
			Name: deploy.config.Namespace,
		}

		namespaceObj := &v1.Namespace{
			ObjectMeta: namespaceObjectMeta,
		}

		namespace, err = deploy.client.CoreV1().Namespaces().Create(context.TODO(), namespaceObj, metav1.CreateOptions{})
		if err != nil {
			return err
		}
		deploy.logger.Infof("Created the %s namespace", deploy.config.Namespace)
	}

	// add extra namespace labels and annotations
	labels := namespace.ObjectMeta.Labels
	if labels == nil {
		labels = make(map[string]string)
	}

	annotations := namespace.ObjectMeta.Annotations
	if annotations == nil {
		annotations = make(map[string]string)
	}

	for key, val := range deploy.config.ExtraNamespaceLabels {
		labels[key] = val
		deploy.logger.Infof("Labeling the %s namespace with '%s=%s'", deploy.config.Namespace, key, val)
	}

	/*
		In the case where the platform is set to Openshift (the default value),
		we need to make a few modifications to the namespace metadata.

		The 'openshift.io/cluster-monitoring' labels tells the cluster-monitoring
		operator to scrape Prometheus metrics for the installed Metering namespace.

		The 'openshift.io/node-selector' annotation is a way to control where Pods
		get scheduled in a specific namespace. If this annotation is set to an empty
		label, that means that Pods for this namespace can be scheduled on any nodes.

		In the case where a cluster administrator has configured a value for the
		defaultNodeSelector field in the cluster's Scheduler object, we need to set
		this namespace annotation in order to avoid a collision with what the user
		has supplied in their MeteringConfig custom resource. This implies that whenever
		a cluster has been configured to schedule Pods using a default node selector,
		those changes must also be propogated to the MeteringConfig custom resource, else
		the Pods in Metering namespace will be scheduled on any available node.
	*/
	if deploy.config.Platform == "openshift" {
		labels["openshift.io/cluster-monitoring"] = "true"
		deploy.logger.Infof("Added the 'openshift.io/cluster-monitoring' label to the %s namespace", deploy.config.Namespace)

		annotations["openshift.io/node-selector"] = ""
		deploy.logger.Infof("Added the empty 'openshift.io/node-selector' annotation to the %s namespace", deploy.config.Namespace)
	}

	namespace.ObjectMeta.Labels = labels
	namespace.ObjectMeta.Annotations = annotations

	_, err = deploy.client.CoreV1().Namespaces().Update(context.TODO(), namespace, metav1.UpdateOptions{})
	if err != nil {
		return fmt.Errorf("failed to update labels and annotations on the %s namespace: %v", deploy.config.Namespace, err)
	}
	return nil
}
```